stage0/src/acpi.rs

```diff
@@ -114,6 +116,118 @@ impl Rsdp {
 
         Ok(())
     }
+
+    pub fn xsdt(&self) -> Result<Option<&Xsdt>, &'static str> {
+        if self.xsdt_address == 0 {
+            Ok(None)
+        } else {
+            Some(Xsdt::new(VirtAddr::new(self.xsdt_address))).transpose()
+        }
+    }
+}
+
+/// Header common for all ACPI tables.
+///
+/// See Section 5.2.6, System Description Table Header, in the ACPI specification for more details.
+#[repr(C, packed)]
+pub struct DescriptionHeader {
+    /// ASCII string representation of the table identifer.
+    pub signature: [u8; 4],
+
+    /// Length of the table, in bytes, including the header.
+    length: u32,
+
+    /// Revision of the struture corresponding to the signature field for this table.
+    revision: u8,
+
+    /// The entire table, including the checksum field, must add to zero to be considered valid.
+    checksum: u8,
+
+    /// OEM-supplied string that identifies the OEM.
+    oem_id: [u8; 6],
+
+    /// OEM-supplied string that the OEM uses to identify the particular data table.
+    oem_table_id: [u8; 8],
+
+    /// OEM-supplied revision number.
+    oem_revision: u32,
+
+    /// Vendor ID of utility that created the table, e.g. the ASL Compiler.
+    creator_id: u32,
+
+    /// Revision of the utility that created the table, e.g. revision of the ASL Compiler.
+    creator_revision: u32,
+}
+
+/// Extended System Description Table.
+///
+/// See Section 5.2.8 in the ACPI specification for more details.
+#[repr(C, packed)]
+pub struct Xsdt {
+    header: DescriptionHeader,
+    // The XSDT contains an array of pointers to other tables, but unfortunately this can't be
+    // expressed in Rust.
+}
+
+impl Xsdt {
+    pub fn new(addr: VirtAddr) -> Result<&'static Xsdt, &'static str> {
+        // Safety: we're checking that the claimed XSDT is entirely within the EBDA.
+        let ebda_base = unsafe { EBDA.as_ptr() } as usize;
+        let xsdt = unsafe { &*(addr.as_ptr() as *const Xsdt) };
+
+        if (addr.as_u64() as usize) < ebda_base
+            || addr.as_u64() as usize + xsdt.header.length as usize > ebda_base + EBDA_SIZE
+        {
+            return Err("XSDT doesn't fit in EBDA");
+        }
+        xsdt.validate()?;
+        Ok(xsdt)
+    }
+
+    pub fn validate(&self) -> Result<(), &'static str> {
+        // Unfortunately we have to exceed our bounds to compute the checksum.
+        // Safety: we've checked that the address + len is weithin the EBDA bounds in `new()`.
```

Comment:
why is this public? is the intention that it may be called outside of the new method? If that is the case, can we sure that "we've checked that the address + len is weithin the EBDA bounds in `new()`" always holds true?